File: src/main/com/natpryce/snodge/mutagens/ReorderObjectProperties.java
package com.natpryce.snodge.mutagens;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.natpryce.snodge.DocumentMutation;
import com.natpryce.snodge.JsonPath;
import com.natpryce.snodge.Mutagen;
import kotlin.jvm.functions.Function1;
import kotlin.sequences.Sequence;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static kotlin.sequences.SequencesKt.emptySequence;
import static kotlin.sequences.SequencesKt.sequenceOf;

public class ReorderObjectProperties implements Mutagen {
    @Override
    public Sequence<DocumentMutation> potentialMutations(JsonElement document, JsonPath pathToElement, JsonElement elementToMutate) {
        if (elementToMutate.isJsonObject()) {
            return sequenceOf(pathToElement.map(this::mutate));
        }
        else {
            return emptySequence();
        }
    }

    public JsonElement mutate(JsonElement element) {
        List<Map.Entry<String, JsonElement>> objectProperties = new ArrayList<>(element.getAsJsonObject().entrySet());
        Collections.shuffle(objectProperties);

        JsonObject mutant = new JsonObject();
        for (Map.Entry<String, JsonElement> property : objectProperties) {
            mutant.add(property.getKey(), property.getValue());
        }
        return mutant;
    }
}

File: src/main/com/natpryce/snodge/mutagens/ReorderObjectProperties.kt
package com.natpryce.snodge.mutagens

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.natpryce.snodge.JsonNodeMutagen
import com.natpryce.snodge.JsonPath
import java.util.ArrayList
import java.util.Collections

class ReorderObjectProperties : JsonNodeMutagen {
    override fun invoke(document: JsonElement, pathToElement: JsonPath, elementToMutate: JsonElement): Sequence<Lazy<JsonElement>> {
        if (elementToMutate.isJsonObject) {
            return sequenceOf(lazy { pathToElement.map(document, this::mutate) })
        }
        else {
            return emptySequence()
        }
    }
    
    fun mutate(element: JsonElement): JsonElement {
        val objectProperties = ArrayList(element.asJsonObject.entrySet())
        Collections.shuffle(objectProperties)
        
        val mutant = JsonObject()
        for ((key, value) in objectProperties) {
            mutant.add(key, value)
        }
        return mutant
    }
}
